govcd/adminvdc.go

```diff
@@ -52,33 +54,273 @@ func (adminVdc *AdminVdc) Refresh() error {
+	apiVersion, err := adminOrg.client.maxSupportedVersion()
+	if err != nil {
+		return nil, err
+	}
+	realFunction, ok := vdcProducerByVersion["vdc"+vcdVersionToApiVersion[apiVersion]]
+	if !ok {
+		return nil, fmt.Errorf("no entity type found %s", "vdc"+apiVersion)
+	}
+	if realFunction.CreateVdc == nil {
+		return nil, fmt.Errorf("function CreateVdc is not defined for %s", "vdc"+apiVersion)
+	}
+
+	util.Logger.Printf("[DEBUG] CreateOrgVdc call function for version %s", realFunction.SupportedVersion)
+	return realFunction.CreateVdc(adminOrg, vdcConfiguration)
+}
+
+// CreateOrgVdcAsync creates a VDC with the given params under the given organization.
+// Returns an Task.
+func (adminOrg *AdminOrg) CreateOrgVdcAsync(vdcConfiguration *types.VdcConfiguration) (Task, error) {
+	apiVersion, err := adminOrg.client.maxSupportedVersion()
+	if err != nil {
+		return Task{}, err
+	}
+	realFunction, ok := vdcProducerByVersion["vdc"+vcdVersionToApiVersion[apiVersion]]
+	if !ok {
+		return Task{}, fmt.Errorf("no entity type found %s", "vdc"+apiVersion)
+	}
+	if realFunction.CreateVdcAsync == nil {
+		return Task{}, fmt.Errorf("function CreateVdcAsync is not defined for %s", "vdc"+apiVersion)
+	}
+
+	util.Logger.Printf("[DEBUG] CreateOrgVdcAsync call function for version %s", realFunction.SupportedVersion)
+
+	return realFunction.CreateVdcAsync(adminOrg, vdcConfiguration)
+}
+
+// createVdc creates a VDC with the given params under the given organization.
+// Returns an Vdc.
+func createVdc(adminOrg *AdminOrg, vdcConfiguration *types.VdcConfiguration) (*Vdc, error) {
+	util.Logger.Printf("[TRACE] createVdc called %#v", *vdcConfiguration)
+	err := adminOrg.CreateVdcWait(vdcConfiguration)
+	if err != nil {
+		return nil, err
+	}
+
+	vdc, err := adminOrg.GetVDCByName(vdcConfiguration.Name, true)
+	if err != nil {
+		return nil, err
+	}
+	return vdc, nil
+}
+
+// updateVdcAsync updates a VDC with the given params. Returns an Task.
+func updateVdcAsync(adminVdc *AdminVdc) (Task, error) {
+	util.Logger.Printf("[TRACE] updateVdcAsync called %#v", *adminVdc)
+	adminVdc.AdminVdc.Xmlns = types.XMLNamespaceVCloud
+
+	// Return the task
+	return adminVdc.client.ExecuteTaskRequest(adminVdc.AdminVdc.HREF, http.MethodPut,
+		types.MimeAdminVDC, "error updating VDC: %s", adminVdc.AdminVdc)
+}
+
+// updateVdc updates a VDC with the given params. Returns an AdminVdc.
+func updateVdc(adminVdc *AdminVdc) (*AdminVdc, error) {
+	util.Logger.Printf("[TRACE] updateVdc called %#v", *adminVdc)
+	task, err := updateVdcAsync(adminVdc)
+	if err != nil {
+		return nil, err
+	}
+
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, err
+	}
 
 	err = adminVdc.Refresh()
 	if err != nil {
-		return AdminVdc{}, err
+		return nil, err
+	}
+
+	return adminVdc, nil
+}
+
+// updateVdcAsyncV97 updates a VDC with the given params. Supports Flex type allocation.
+// Needs vCD 9.7 to work. Returns an Task.
```

Comment:
Looking at the code below I think this is the correct version notation:
```suggestion
// Needs vCD 9.7+ to work. Returns an Task.
```

Reply:
fixed